### PR TITLE
Allow filtering in Scenery window

### PR DESCRIPTION
**Pull request by trzejos**
This change adds a text box and clear button between the tabs and scenery list in the Scenery window.

Typing in the text box allows the user to filter the objects in the current tab based on the following criteria:
- Name `Object::GetName()`
- Authors `Object::GetAuthors()`
- Identifier `Object::Identifier()`
- Filename `ObjectRepositoryItem::Path`

This works by creating a new list of `ScenerySelection` objects, which is used to make a copy of the current tab's entry list with the appropriate objects filtered out.

When no objects are available with the current selection, the tool is set to `Tool::Arrow` and placing scenery is not allowed. Right clicking still will delete scenery/paths in this state.

Todo:
- [x] Use ObjectManager to get object info
- [x] Clear scenery selection when switching to an empty tab
- [x] Fix bug with scenery tool when an undefined selection is made

Potential ideas for another PR:
- Have a unique filter string for each tab, rather than a single string for all tabs
- Add an additional tab after misc for all scenery, allowing users to search all scenery objects in one view

Examples:
image
image
image

### CODE REVIEW

File: src/openrct2/world/Scenery.h [hunk 1]
```diff
@@ -28,6 +29,8 @@ struct SceneryEntryBase
 {
     StringId name;
     uint32_t image;
+    SceneryObject* obj;
```
Comment (duncanspumpkin) [inline src/openrct2/world/Scenery.h:32]: There is no need to add this pointer you can get the name through other means. I'm about to remove this struct as well.

File: src/openrct2-ui/windows/Scenery.cpp [hunk 2]
```diff
@@ -997,6 +1043,77 @@ class SceneryWindow final : public Window
...
+    }
+
+    bool IsFiltered(const ScenerySelection& selection)
+    {
+        auto empty = _filter.empty();
+        auto length = _filter.length();
+        if (empty || length == 0)
+            return true;
+
+        const SceneryEntryBase* sceneryEntry = nullptr;
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/Scenery.cpp:1065]: You should get the object not the entry. Use `Context::GetObjectManager::GetLoadedObject`

File: src/openrct2/world/Scenery.h (hunk 1 above)
Comment (trzejos) [reply src/openrct2/world/Scenery.h]: This has been reverted

File: src/openrct2-ui/windows/Scenery.cpp (hunk 2 above)
Comment (trzejos) [reply src/openrct2-ui/windows/Scenery.cpp]: done

**Review by duncanspumpkin — COMMENTED** [review]
Code all looks sensible. Just need to give it a test now.

**Review by duncanspumpkin — APPROVED** [review]
@Broxzier I just now tested it and it appears to work.

**Comment by Broxzier** [comment]
Yeah, I'm sure the searching works as expected. I wanted to test if scroll positions still work well, especially if you switch between tabs and modify the search terms before going back.

**Review by Broxzier — APPROVED** [review]
One thing I don't like with this new feature is that the currently selected object will change when it doesn't match the filter. I think that for each tab, the selected object should remain selected, and only change when the player actually clicks on another object.

This can be done later, however. This is already a very welcome feature as is.

**Comment by trzejos** [comment]
Yeah I implemented it that way on purpose. It felt weird to me being able to place a scenery object when you can't see it in the scenery window. We might be able to get around this by putting the current selection in the filtered list if it already isn't present instead of clearing it. That would make it so your selected scenery always shows up even when it doesn't match the filter though.

**Pull request merged**